File: scripts/xcheck_rfft.py
#!/usr/bin/env python3
"""Cross-validate pan's rfftForward against scipy.fft.rfft (on-demand, NOT hermetic).
Reads pan's bins from stdin ('re im' per line) and compares to scipy on the same
deterministic analytic signal. Exits 0 iff allclose."""
import sys, numpy as np
from scipy.fft import rfft
N = 1024
n = np.arange(N)
x = np.sin(2*np.pi*3*n/N) + 0.5*np.sin(2*np.pi*7*n/N)
ref = rfft(x)  # length N/2+1 complex
pan = []
for line in sys.stdin:
    re, im = line.split()
    pan.append(float(re) + 1j*float(im))
pan = np.array(pan)
ok = np.allclose(pan, ref, atol=1e-2, rtol=1e-3)
maxerr = np.max(np.abs(pan - ref))
print(f"pan rfft vs scipy.fft.rfft: allclose={ok}  max_abs_err={maxerr:.6g}  bins={len(pan)}")
sys.exit(0 if ok else 1)

File: scripts/xcheck_rfft.zig
//! On-demand cross-validation: dump pan's rfftForward of a deterministic analytic
//! signal (re/im per line to stderr), compared to scipy.fft.rfft by xcheck_rfft.py.
//! NOT part of `zig build test` (the hermetic suite uses an in-test naive DFT).
const std = @import("std");
const pan = @import("pan");
pub fn main() void {
    const N = 1024;
    var x: [N]f32 = undefined;
    for (&x, 0..) |*v, n| {
        const t: f32 = @floatFromInt(n);
        v.* = @sin(2.0 * std.math.pi * 3.0 * t / N) + 0.5 * @sin(2.0 * std.math.pi * 7.0 * t / N);
    }
    var bins: [N / 2 + 1]std.math.Complex(f32) = undefined;
    pan.spectral.rfftForward(f32, N, &x, &bins);
    for (bins) |c| std.debug.print("{d} {d}\n", .{ c.re, c.im });
}
